unit test for flux_dataset dataloader checkpointing

wesleytruong opened the pull request:
Adds a unit test for loading flux dataset from a checkpoint. Creates the new dataloader from a checkpoint then ensures that the next generated labels and tokens are the same in both dataloaders, starting from the checkpoint.

wwwjn commented on torchtitan/experiments/flux/tests/unit_tests/test_flux_dataloader.py:
```diff
@@ -97,3 +99,63 @@ def _build_dataloader(
+                config = config_manager.parse_args(
+                    [
+                        f"--experimental.custom_args_module={path}",
+                        # Profiling options
```
If these options are not needed, pls delete them, I forgot to clean them in `_test_flux_dataloader`

wwwjn commented on torchtitan/experiments/flux/tests/unit_tests/test_flux_dataloader.py:
```diff
@@ -97,3 +99,63 @@ def _build_dataloader(
+                        "--training.seed",
+                        "0",
+                        "--training.classifer_free_guidance_prob",
+                        "0.447",
```
Why enabling classifer-free guidance here?  Would it cause random dropout in the dataloader?

wwwjn commented on torchtitan/experiments/flux/tests/unit_tests/test_flux_dataloader.py:
```diff
@@ -97,3 +99,63 @@ def _build_dataloader(
+                for i in range(50):
+                    torch.manual_seed(i)
+                    expected_input_ids, expected_labels = next(it)
+                    torch.manual_seed(i)
```
You could add some NOTE here, saying that we need to manually set random seed every time before calling next() because the randomness in dataloader behaviour.

wwwjn reviewed (commented):
Nice work! Please fix lint as well

tianyu-l commented on torchtitan/experiments/flux/tests/unit_tests/test_flux_dataloader.py:
```diff
@@ -16,6 +18,7 @@ class TestFluxDataLoader:
     def test_load_dataset(self):
         for dataset_name in ["cc12m-test"]:
             self._test_flux_dataloader(dataset_name)
+            self._test_flux_dataloader_checkpointing(dataset_name)
```
With the new one, the old one's value is not clear.
I suggest we merge the two into one.
Also there is no need for this for-loop and extra layer of `test_load_dataset` -> `_test_flux_dataloader`, we can do the test just in one function.

tianyu-l commented on torchtitan/experiments/flux/tests/unit_tests/test_flux_dataloader.py:
```diff
@@ -97,3 +93,55 @@ def _build_dataloader(
             tokenizer=None,
             infinite=True,
         )
```
`_build_dataloader` on line 87 is not very useful, we can just directly use `build_flux_dataloader`

wesleytruong commented:
Refactored the two tests according to @tianyu-l
The test now checks for the correct tensor shapes during the first num_steps, then for the next num_steps it ensures the loaded from checkpoint dataloader generates tokens and labels correctly
```build_dataloader``` is also called directly now instead of through ```_build_dataloader```

tianyu-l commented on torchtitan/experiments/flux/tests/unit_tests/test_flux_dataloader.py:
```diff
@@ -14,86 +16,99 @@
+                    ]
+                )
+
+                with maybe_enable_profiling(
+                    config, global_step=0
+                ) as torch_profiler, maybe_enable_memory_snapshot(
+                    config, global_step=0
+                ) as memory_profiler:
```
I think we can remove these code related to profiling.

tianyu-l commented on torchtitan/experiments/flux/tests/unit_tests/test_flux_dataloader.py:
```diff
@@ -14,86 +16,99 @@
+                        torch.manual_seed(i)
+                        input_ids, labels = next(it_resumed)
+
+                        if torch_profiler:
+                            torch_profiler.step()
+                        if memory_profiler:
+                            memory_profiler.step()
```
remove

tianyu-l commented on torchtitan/experiments/flux/tests/unit_tests/test_flux_dataloader.py:
```diff
@@ -14,86 +16,99 @@
+                        )
+                        assert torch.equal(labels, expected_labels)
+
+                    if torch_profiler:
+                        torch_profiler.step()
+                    if memory_profiler:
+                        memory_profiler.step(exit_ctx=True)
```
remove

tianyu-l reviewed (commented):
Looks better. I had one more comment on removing profiling code, as they are not quite related to the test itself.

> The test now checks for the correct tensor shapes during the first num_steps, then for the next num_steps it ensures the loaded from checkpoint dataloader generates tokens and labels correctly

This is helpful. Let's put them as comments in the test itself, so that people who look at the test later can easily understand what you are doing here.

wesleytruong commented:
Removed the memory profiling, and added the unit test for tensor dimensions. @tianyu-l 
Added unit the test for the tensor shape after tokenization, but used a different value for the test. I believe this is the correct shape since it matches the shape of a tokenized sequence. @wwwjn
```
                    assert input_data["clip_tokens"].shape == (
                        batch_size,
                        1,
                        77,
                    )
                    assert input_data["t5_tokens"].shape == (
                        batch_size,
                        1,
                        256,
                    )
```

tianyu-l reviewed (approved):
Looks great, thank you!
